src/db/pool.cr

```diff
@@ -47,6 +60,56 @@ module DB
       @initial_pool_size.times { build_resource }
     end
 
+    def reap_connection!(connection : T, delay : Float64)
+      return unless connection.responds_to?(:check)
+
+      sleep delay
+
+      sync do
+        connection_idle = @idle.includes? connection
+        return unless connection_idle # let the next reaping attempt get it
+
+        delete connection
+      end
+
+      connection.check
+
+      sync do
+        @total << connection
+        @idle << connection
+      end
+    rescue exc : DB::ConnectionLost
+      connection.close
+
+      sync do
+        add_resource! if can_increase_pool?
+      end
+    end
+
+    def add_resource!
+      @inflight += 1
+      r = unsync { build_resource }
+    ensure # prevent inflight from incrementing if build_resource fails
+      @inflight -= 1
+      r
+    end
+
+    def start_reaper!
+      spawn do
+        sleep @reaping_delay
+
+        loop do
+          sync do
+            @idle.each_with_index do |connection, i|
+              spawn reap_connection!(connection, i * 0.1)
```

Comment:
The idea with the 100ms stacking delay here is to prevent clearing out all of the `@idle` connections at the same time.

Reply:
Could we check the individual connections consecutively instead of concurrently?
Spawning a new fiber for each connection seems unnecessary.

Reply:
Actually, every fiber acquires a lock in `reap_connection!`, so they would execute consecutively anyways.

Reply:
I think this spawn is just a form of creating a timer here

Reply:
Yeah, this was intended to be a timer.

My goal here was to create a timer that injected 100ms of time in between attempts to remove an `@idle` connection -- the goal here was to attempt to make sure that the `@idle` pool always retains a spare connection (aside from when using `max_idle_pool_size=1`).

Reply:
Sure. But your code essentially executes equivalent to this plain flow. 

```cr
@idle.each_with_index do |connection, i|
  reap_connection!(connection, i * 0.1)
end
```

The fibers don't actually run concurrently because of the mutex. That also wrecks the delay. All fibers sleep as soon as they are spawned, then they block waiting for the mutex (except the first one). When the previous one releases the mutex, the next executes directly because it's already waiting. So the delay is without effect.

Reply:
That makes sense, thanks.

Given that, what do you think about something like this? The main thing I'm not certain of is whether the health check should live inside of the mutex (my guess is these should usually resolve in ~100ms, and aside from PG, not many adapters may choose to implement this) -- if so, the implementation can be really clean, as the health check itself would act as the delay:

```crystal
    def start_reaper!
      spawn do
        sleep @reaping_delay

        loop do
          @idle.each { |connection| reap_connection! connection }
          sleep @reaping_frequency
        end
      end
    end
    
    def reap_connection!(connection : T)
      sync do
        if @idle.includes? connection
          delete connection
          connection.check
          @total << connection
          @idle << connection
        end
      end
    rescue exc : PoolResourceLost(T)
      connection.close
      sync { add_resource! if can_increase_pool? }
    end

```

Reply:
I don't think the health check should be synced. It includes an network request, hence there is no guarantee how long it might take to complete.

Reply:
Makes sense -- updated accordingly.